src/libPMacc/include/particles/operations/CountParticles.hpp

```diff
@@ -31,72 +30,163 @@
 #include "particles/particleFilter/PositionFilter.hpp"
 #include "nvidia/atomic.hpp"
 #include "memory/shared/Allocate.hpp"
-
+#include "traits/GetNumWorkers.hpp"
+#include "mappings/threads/ForEachIdx.hpp"
+#include "mappings/threads/IdxConfig.hpp"
 
 
 namespace PMacc
 {
 
-/* count particles in an area
- * is not optimized, it checks any partcile position if its realy a particle
+/* count particles
+ *
+ * it is allowed to call this kernel on frames with holes (without calling fillAllGAps before)
+ *
+ * @tparam T_numWorkers number of workers
  */
+template< uint32_t T_numWorkers >
 struct KernelCountParticles
 {
-    template<class PBox, class Filter, class Mapping>
-    DINLINE void operator()(
-        PBox pb,
+    /** count particles
+     *
+     * @tparam T_PBox PMacc::ParticlesBox, particle box type
+     * @tparam T_Filter functor to filter particles
+     * @tparam T_Mapping supercell mapper functor type
+     *
+     * @param pb particle memory
+     * @param gCounter pointer for the result
+     * @param filter functor to filter particles those should be counted
+     * @param mapper functor to map a block to a supercell
+     */
+    template<
+        typename T_PBox,
+        typename T_Filter,
+        typename T_Mapping
+    >
+    DINLINE void operator( )(
+        T_PBox & pb,
         uint64_cu* gCounter,
-        Filter filter,
-        Mapping mapper
+        T_Filter & filter,
+        T_Mapping & mapper
     ) const
     {
+        using namespace mappings::threads;
 
-        typedef typename PBox::FrameType FRAME;
-        typedef typename PBox::FramePtr FramePtr;
-        const uint32_t Dim = Mapping::Dim;
+        using Frame = typename T_PBox::FrameType;
+        using FramePtr = typename T_PBox::FramePtr;
+        constexpr uint32_t dim = T_Mapping::Dim;
+        constexpr uint32_t frameSize = math::CT::volume< typename Frame::SuperCellSize >::type::value;
+        constexpr uint32_t numWorkers = T_numWorkers;
 
-        PMACC_SMEM( frame, FramePtr );
-        PMACC_SMEM( counter, int );
-        PMACC_SMEM( particlesInSuperCell, lcellId_t );
+        PMACC_SMEM(
```

Comment:
its quite weird to write this `__shared__ frame = FramePtr;` assignment in such a way...
can't we make it
```C++
auto frame = PMACC_SMEM( FramePtr );
```
? :D